Add Semtech Lora radio drivers

kivaisan opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Add Semtech Lora radio drivers for SX1272, SX1276 and SX126x radios.
Added as subtree from original repository: https://github.com/ARMmbed/mbed-semtech-lora-rf-drivers

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

Enable using Lora with no additional repositories, all of them part of Mbed OS.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

None

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [X] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
Tests run with our internal Lora test suite

----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@ARMmbed/mbed-os-wan @kjbracey-arm 

----------------------------------------------------------------------------------------------------------------

Commit:
Initial commit

Commit:
Removing namespace pollution & rf ctrls refactor

rf_ctrls is a data structure that holds all control pins for a certain LoRa Radio.
Same data structure was being used in both SX1272 and SX1276 drivers. As we are
heading towards putting both drivers in the same repo and as the plan is to use them
togather, i.e., both header files could be included we need to move this data structure to a
common location so that the name does not collide.

In addition to that we have tried to reduce namespace pollution caused by the driver
header files.

Commit:
Merge pull request #1 from ARMmbed/ns_pollution

Removing namespace pollution & rf ctrls refactor

Commit:
Buffer Size and missing terminator for SX127 class

For some unknown reason (may be a rebase issue) Mbed config name for the
drivers was not spell correctly inside the source macro definition.
Alongwith that the SX1276 driver was missing a class terminator.

Commit:
Merge pull request #2 from ARMmbed/bug_fix

Buffer Size and missing terminator for SX127 class

Commit:
[IOTCELL-284] Removing NULL checks from Callbacks

In respose to the radio event callback change from c style callbacks
to Mbed callbacks, we need to remove NULL checks from the driver code
as the callbacks itself are not pointers anymore. However, thanks to
template magic, we can check if callback is assigned or not. If its not
not assigned, it is default constructed to NULL which we can check by boolean
operator.

Commit:
Merge pull request #3 from ARMmbed/radio_event_callbacks

[IOTCELL-284] Removing NULL checks from Callbacks

Commit:
Add support for Multitech xDot module

xDot module does not have external antenna control pins nor DIO5
pin connected.

Commit:
Enable PA_BOOST for xDOT module

Commit:
Use PA_BOOST for WISE-1510 module

Commit:
Set up DIO4 and 5 interrupts only if pins are connected

For example muRata modules do not have DIO4 and DIO5 pins connected.
If these pins are not connected, driver should not try to set up
interrupt handlers for those.

Commit:
Add MTB_MTS_XDOT target

Commit:
Change WISE-1510 target to new name

Commit:
Fixing paths to LoRaRadio and warnings

LoRaRadio is now moved to feature/lorawan from feature/netsocket.

Commit:
Enable TCXO always if pin is defined

Commit:
Add TCXO pin for SX1272

Also reformatted constructor params list to more readable format

Commit:
Checking for pins being NC before writing

NXP Hal provides a write API on DigitalOut::write() that asserts if pin
is NC and we try to perform an operation on it. This behaviour is not
consistent among various Hals in Mbed-OS. However, we now check the pin
ourselves in the driver just like 1276.

Commit:
Fix IAR compiler warnings

preamble_detected and sync_word_detected are uint8_t integer type variables so those should not be compared to (boolean) true/false values.

Fixed to use 1 and 0 values instead.

Commit:
Merge pull request #12 from ARMmbed/fix_iar_compiler_warnings

Fix IAR compiler warnings

Commit:
Added flagging for DEVICE_SPI

Commit:
Remove product specific variables and target flags

Since variant detection is using ant_switch pin, it is now only used if that pin
has been connected. Otherwise generic variant code is used.

Commit:
Change static variables to class members

Some variables were still defind as static in sources but as drivers
are now C++ classes, these variables should be members of the class.

Commit:
Remove usage of DEVICE_SPI flag

DEVICE_SPI flag is no longer needed as our CI only builds
mbed-os-example-lorawan to targets which really have lora hw.

Commit:
Remove obsolete FEATURE_COMMON_PAL flag

mbed-trace no longer requires COMMON_PAL.

Commit:
Fix correct pin check for rxctl

Fixes issue #20

Commit:
Removing software RX timeouts in RX chain

Both in FSK and LoRa mode, we do not rely on software timers anymore.
This makes our timing precise and enables us to consume less power.
It also gives us flexibility which we need for future deep sleep
venture as Mbed OS timeout class may lock the sleep manager if not using
low power timers.

Commit:
Changing double precision to float

We shouldn't use double precision as most of te embedded processors may
not support it.

Commit:
Deprecating receive(uint32_t) API

We are deprecating receive(uint32_t) API in favour of receive(void)
API because we are ditching software timeout timers in the driver
for RX chain.

Commit:
Adding assert if rx timeout is out of range

Rx timeout in symbols cannot become more than 255 as the preamble length
is fixed. For diagonostics purposes we add an MBED_ASSERT.

Commit:
TX timeout issue fix

In case of catastrophic bus failure at radio end, we shouldn't reset the chip
as it will cause the chip to hang and then burst out interrupts in a fury resulting
in an ISR queue overflow.
rather than that we gracefully accept the failure and set the radio to sleep and
set the state to idle. In addition to that we inform the upper layers about the
failure.

A little touch up to the FHSS case, was actually a leftover from the previous PR.
We don't use FHSS mode, that was why we didn't catch in the testing.

Commit:
Correcting buffer size and tx timer units

This commit fixes two issues reported in #24 and #25.
The default buffer size for the radio driver should have been 255 (0xFF) not 256.
In addition to that it was pointed out that the tx timer values were inconsistent
with the doc. All timer values are considered in ms except symbol timeout and the
time value for set_tx_continuous_wave() API.

mattbrown015 committed:
Allow configuration of the radio variant

mattbrown015 committed:
Corrected 'help' in SX1272 config item

Commit:
Adding support for low power timeouts

We had been on a mission to get rid of all the Timeout instances from
the radio drivers. However, 'tx_timeout' is a necessary evil which runs
every time we transmit so that we can inform the stack about any fatal
errors. The side effect of this Timeout is that it was using high-res
timeout which would lock the deep sleep and hence prevent the MCU to go
to deep sleep.

We have now added support for low power timeout and we check compile
time if the platform supports low power timeout to use it in the driver.
This means that now the 'tx_timeout' will not lock the deep sleep.

Commit:
Using new rtos::Thread APIs

Some of the rtos::Thread APIs like signal_set() and signal_wait() are
noe deprecated. Although they are still backward compatible, we switch
to the new APIs.

Commit:
Adding SX126X radio driver

Driver compatible with SX126X series. Operates by default with Cold
Start sleep mode, i.e., the lowest possible power state while sleeping.
RX boost mode supported and the driver can be configured to get RX gain
at the orice of power. Default mode is normal RX gain.

Commit:
Removing mbed.h inclusion

It's better to include what we need rather than whole mbed.h.
Small style knits.

Commit:
Corrected copyright for sx126x_ds.h

Commit:
Fix SPI flags for SX1272 and SX1276

Commit:
Lora: Replace deprecated methods with new ones in Semtech drivers

Commit:
Astyle fixes

Commit:
BUGFIX: MODEM_LORA rx continuous was timing out

Commit:
set symb_timeout to 0 also in FSK mode

Commit:
Add 'components/lora/' from commit '0f7efe3ad17c6c619843b8391b3035149f358260'

git-subtree-dir: components/lora
git-subtree-mainline: a6ef9db8dc259f6f1f042e1009ec8920769af1c7
git-subtree-split: 0f7efe3ad17c6c619843b8391b3035149f358260

Commit:
Lora: Fix line endings

AnttiKauppila reviewed (changes requested):
Internal check ongoing

AnttiKauppila reviewed (dismissed):
Ok to proceed

Commit:
Lora: Add driver implementation inside DEVICE_SPI flag

These drivers uses SPI to communicate with radio so SPI must be present
when compiling these.

Commit:
Lora: Move drivers to own COMPONENT -directories

To enable building only when driver is enabled with components_add -setting
in json configuration.
